### ISSUE TITLE
Come up with a policy for bumping toolchains

### ISSUE THREAD

**Issue by stevenhorsman** [kata-containers/kata-containers#9841]
In the AC meeting on 6th June 2024 (https://etherpad.opendev.org/p/Kata_Containers_2024_Architecture_Committee_Mtgs#L45) we had a discussion about the balance of keeping our toolchains up-to-date (and the crates/packages used regularly updated) to drive out potential vulnerabilities and concerns that the distro team might have about regularly bumping.

In order to provide clarity there was a suggestion to come up with a policy for our toolchain changes and document that, so that all parties understand where they stand in respect to this.

For example we could say that every odd release we will bump to the latest stable version of golang and rust and leave it unchanged in between, or say we will bump versions every 6 months etc.

**Comment by beraldoleal** [comment]
Hi @stevenhorsman ...

> For example we could say that every odd release we will bump to the latest stable version of golang and rust and leave it unchanged in between, or say we will bump versions every 6 months etc.

IIUC, the idea is to bump on *minor* odd releases. Is that right?

If so, my concern is that we might (unintentionally) break behaviour by bumping tooling during minor releases.

> When MINOR increases, the new release adds new features but without changing the existing behavior.

**Comment by stevenhorsman** [comment]
@beraldoleal - I was just listing an example of something we could do, rather than a community proposal. Nevertheless, I don't think there is likely to be a change of behaviour, other that removing security bugs, when bumping the toolchain.

If your proposal is that we only bump toolchains on major releases, then either we need a lot more major releases, or we wrap up the project as using two year old toolchains that are past EoL by nearly two years (rust for example only has 6 weeks before the next release makes it EoL) means we are going to be broken with CVE and not able to get new features from dependencies, making development impossible.

**Comment by beraldoleal** [comment]
> @beraldoleal - I was just listing an example of something we could do, rather than a community proposal. Nevertheless, I don't think there is likely to be a change of behaviour, other that removing security bugs, when bumping the toolchain.
> 
> If your proposal is that we only bump toolchains on major releases, then either we need a lot more major releases, or we wrap up the project as using two year old toolchains that are past EoL by nearly two years (rust for example only has 6 weeks before the next release makes it EoL) means we are going to be broken with CVE and not able to get new features from dependencies, making development impossible.

Yes, I agree with you, we need to find a middle ground. I was just pointing to our MAJOR vs MINOR "agreement", that we should just take that into consideration. I believe this will be case by case, tho.

**Comment by stevenhorsman** [comment]
There was a discussion (rust specific) in CoCo call today. As part of this Mikko shared the cloud hypervisor approach: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/0fbba66b21afe31ba191d531b36a7dd98750a6ad/Cargo.toml#L11-L17

> ``` # Policy on MSRV (see #4318):
> # Can only be bumped by:
> # a.) A dependency requires it,
> # b.) If we want to use a new feature and that MSRV is at least 6 months old,
> # c.) There is a security issue that is addressed by the toolchain update.
> ```

**Comment by stevenhorsman** [comment]
We have another situation with https://github.com/kata-containers/kata-containers/pull/11018 and the go runtime where a CVE fix needs a bump to go 1.23, but some downstream projects are limited to golang 1.22 still.

**Comment by stevenhorsman** [comment]
Another thing we could do that is related to this is try and build our components on both the MSRV (like we currently do) and the stable version, so we can find and get ahead of potential issues. There are definitely some challenges with that as I found with the 1.80 bump that some changes we needed were only available in 1.80, so we couldn't use them until that was our MSRV, but it might help us prepare for the upgrade when the time comes.

**Comment by stevenhorsman** [comment]
We discussed this on the vPTG. My understanding of the general consensus is that
- we should bump the golang "major" release every six months, to line up with golang releases and ensure that we are on the oldest of the two support releases unless there is a strong reason to move to a newer version. Minor releases should be bumped as security vulnerabilities require
- Rust is more complicated. The cloud-hypervisor approach of:
> MSRV Can only be bumped by:
> a.) A dependency requires it,
> b.) If we want to use a new feature and that MSRV is at least 6 months old,
> c.) There is a security issue that is addressed by the toolchain update.

seems like reasonable starting point as we try and find the balance between supporting downstream consumers and not being blocked with features and especially vulnerabilities by just one downstream.
